chain/events/filter/index.go

```diff
@@ -56,27 +61,203 @@ var ddls = []string{
+
+			err2 := tx.Rollback()
+			if err2 != nil {
+				log.Errorf("error in rollback: %s", err2)
+			}
+		}
+	}()
+
+	stmtDeleteOffChainEvent, err := tx.Prepare("DELETE FROM event WHERE tipset_key_cid!=? and height=?")
+	if err != nil {
+		return xerrors.Errorf("prepare stmtDeleteOffChainEvent: %w", err)
+	}
+
+	stmtSelectEvent, err := tx.Prepare("SELECT id FROM event WHERE tipset_key_cid=? ORDER BY message_index ASC, event_index ASC, id DESC LIMIT 1")
+	if err != nil {
+		return xerrors.Errorf("prepare stmtSelectEvent: %w", err)
+	}
+
+	stmtDeleteEvent, err := tx.Prepare("DELETE FROM event WHERE tipset_key_cid=? AND id<?")
+	if err != nil {
+		return xerrors.Errorf("prepare stmtDeleteEvent: %w", err)
+	}
+
+	// get the lowest height tipset
+	var minHeight sql.NullInt64
+	err = ei.db.QueryRow("SELECT MIN(height) FROM event").Scan(&minHeight)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return nil
+		}
+
+		return xerrors.Errorf("query min height: %w", err)
+	}
+
+	currTs := chainStore.GetHeaviestTipSet()
+	log.Infof("Migrating events from head to %d", minHeight.Int64)
+
+	for {
+		if int64(currTs.Height()) < minHeight.Int64 {
+			break
+		}
+
+		tsKey := currTs.Parents()
+		var ts *types.TipSet
+		ts, err = chainStore.GetTipSetFromKey(ctx, tsKey)
+		if err != nil {
+			return xerrors.Errorf("get tipset from key: %w", err)
+		}
+		log.Debugf("Migrating height %d\n", ts.Height())
+
+		var tsKeyCid cid.Cid
+		tsKeyCid, err = ts.Key().Cid()
+		if err != nil {
+			return fmt.Errorf("tipset key cid: %w", err)
+		}
+
+		// delete all events that are not in the canonical chain
+		_, err = stmtDeleteOffChainEvent.Exec(tsKeyCid.Bytes(), ts.Height())
+		if err != nil {
+			return xerrors.Errorf("delete off chain event: %w", err)
+		}
+
+		// find the first eventId from the last time the tipset was applied
+		var eventId sql.NullInt64
+		err = stmtSelectEvent.QueryRow(tsKeyCid.Bytes()).Scan(&eventId)
+		if err != nil {
+			if err == sql.ErrNoRows {
+				currTs = ts
+				continue
+			}
+			return xerrors.Errorf("select event: %w", err)
+		}
+
+		// this tipset might not have any events which is ok
+		if !eventId.Valid {
+			currTs = ts
+			continue
+		}
+		log.Debugf("Deleting all events with id < %d at height %d\n", eventId.Int64, ts.Height())
+
+		var res sql.Result
+		res, err = stmtDeleteEvent.Exec(tsKeyCid.Bytes(), eventId.Int64)
+		if err != nil {
+			return xerrors.Errorf("delete event: %w", err)
+		}
+
+		var nrRowsAffected int64
+		nrRowsAffected, err = res.RowsAffected()
+		if err != nil {
+			return xerrors.Errorf("rows affected: %w", err)
+		}
+		log.Debugf("deleted %d events from tipset %s\n", nrRowsAffected, tsKeyCid.String())
+
+		currTs = ts
+	}
+
+	// delete all entries that have an event_id that doesn't exist (since we don't have a foreign
+	// key constraint that gives us cascading deletes)
+	var res sql.Result
+	res, err = tx.Exec("DELETE FROM event_entry WHERE event_id NOT IN (SELECT id FROM event)")
```

Comment:
In general, avoid `var` unless it actually simplifies things. Here, we can just shadow `err`.

Reply:
I am doing this intentionally as to not shadow `err` so the defer block that runs in the end will detect whether it needs to rollback the transaction or not. If I shadow it, then it it won't work. Explicitly rollbacking the transaction on error is a good practice to reduce locks held, minimize the time transaction is open, etc.

Reply:
Ah, ok. So, for the defer block, I'd:

1. Name the returned error something like `_reterr`.
2. Check `_reterr` in the defer block.

Reply:
E.g., see https://github.com/filecoin-project/lotus/blob/8b849e6ad983c4020b94e582ca645d90c4874830/cmd/lotus-sim/simulation/simulation.go#L184-L199

Reply:
Oh, cool, I didnt know that was possible

Reply:
Alternatively we can just call `defer tx.Rollback()` as soon as we start the transaction. If we committed the transaction, then that calling Rollback is a no-op, and otherwise we rollback. Then we don't need to have named returns

Reply:
Yeah, that's definitely the best approach. Named returns are... best avoided unless you really need them.